lib/granite/action/instrumentation.rb

```diff
@@ -0,0 +1,25 @@
+module Granite
+  class Action
+    module Instrumentation
+      def perform!(*, **)
+        instrument_perform(:perform!) { super }
+      end
+
+      def perform(*, **)
+        instrument_perform(:perform) { super }
+      end
+
+      def try_perform!(*, **)
+        instrument_perform(:try_perform!) { super }
+      end
+
+      private
+
+      def instrument_perform(using)
+        ActiveSupport::Notifications.instrument('granite.perform_action', action: self, using: using) do
```

Comment:
@andriusch Why are we passing `self`?

Reply:
Not sure I understand the question correctly, but basically we need to know which action was performed when `granite.perform_action` occurs. Since `ActiveSupport::Notifications` doesn't pass self by default (given that it's a singleton class method it doesn't have access to `self`), I just pass it as payload to the event.

Reply:
@andriusch What I mean here: what do we need from the entire Granite Action instance (aka `self`)? Do we need its name? Do we need a deeper introspection in its state? Is it passed because the listener can manipulate the action?

Reply:
We definitely need a deeper inspection into action's state, I don't expect actions to be manipulated in notifications, though I wouldn't count that out, notifications can be used essentially as middleware, so it's hard to say at this point what this will or won't be used. I do think though that given event `granite.perform_action`, I would expect access to action that is being performed.

Reply:
@jodosha if you haven't seen it ^

Reply:
That object with large interface is already exposed to outside world, because someone in that world has to call `perform!` on this action. 

I also don't understand why it would be harder to revert? If we want to revert it the difficulty will be the same whether we pass `self` or identifier. Same thing with changing attribute, it will be the same difficulty and I really don't see a reason why notification regarding that action was performed wouldn't have a reference to the action that was performed.

Not to mention that identifier isn't enough to implement the RSpec matcher. And even if it was a specific application could require more things to match in the matcher.